### HYPOTHESIS TESTING

In [48]:
import pandas as pd

In [113]:
!wget -O Final_Total_Dataframe_Stats.csv "https://raw.githubusercontent.com/SayakKDutta1/MSEPROJECT/main/Final_Total_Dataframe_Stats.csv" #https://www.eia.gov/electricity/data/state/  # it gets updated every month in website we have taken till december 2023
df = pd.read_csv('Final_Total_Dataframe_Stats.csv')

--2024-04-30 17:54:39--  https://raw.githubusercontent.com/SayakKDutta1/MSEPROJECT/main/Final_Total_Dataframe_Stats.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3578580 (3.4M) [text/plain]
Saving to: ‘Final_Total_Dataframe_Stats.csv’

Final_Total_Datafra 100%[===================>]   3.41M  --.-KB/s    in 0.09s   

2024-04-30 17:54:39 (36.8 MB/s) - ‘Final_Total_Dataframe_Stats.csv’ saved [3578580/3578580]



In [50]:
df.head()

Unnamed: 0 State RESIDENTIAL (Revenue) RESIDENTIAL (Sales)  \
0           0    AK                36,139            2,22,134   
1           1    AL              2,76,749           28,18,685   
2           2    AR              1,25,142           14,48,385   
3           3    AZ              2,21,494           22,84,523   
4           4    CA             11,87,506           80,64,319   

  RESIDENTIAL (Customers)  RESIDENTIAL (Price) COMMERICIAL (Revenue)  \
0                2,70,456                16.27                36,037   
1               21,32,487                 9.82              1,66,408   
2               13,17,936                 8.64                67,431   
3               25,63,532                 9.70              1,85,595   
4             1,33,52,574                14.73             11,34,995   

  COMMERICIAL (Sales) COMMERICIAL (Customers) COMMERICIAL (Price)  ...  \
0            2,56,270                  47,334               14.06  ...   
1           16,79,995                3,51,480                9.91  ...   
2            8,89,269                1,78,058                7.58  ...   
3           21,56,813                2,99,584                8.61  ...   
4         1,00,31,863               18,58,683               11.31  ...   

  TRANSPORTATION (Revenue) TRANSPORTATION (Sales) TRANSPORTATION (Customers)  \
0                        0                      0                          0   
1                        0                      0                          0   
2                        4                     39                          2   
3                        0                      0                          0   
4                    5,614                 68,156                        495   

  TRANSPORTATION (Price) TOTAL (Revenue) TOTAL (Sales) TOTAL (Customers)  \
0                      0          87,685      5,94,006          3,19,030   
1                      0        6,03,260     69,64,800         24,93,148   
2                  10.26        2,68,283     36,21,183         15,25,714   
3                      0        4,62,196     53,41,392         28,70,201   
4                   8.24       26,79,317   2,19,42,679       1,52,86,390   

  TOTAL (Price)        Date  STATE_NAME  
0         14.76  2009-12-01      Alaska  
1          8.66  2009-12-01     Alabama  
2          7.41  2009-12-01    Arkansas  
3          8.65  2009-12-01     Arizona  
4         12.21  2009-12-01  California  

[5 rows x 24 columns]

In [51]:
df.isnull().sum()

Unnamed: 0                    0
State                         0
RESIDENTIAL (Revenue)         0
RESIDENTIAL (Sales)           0
RESIDENTIAL (Customers)       0
RESIDENTIAL (Price)           0
COMMERICIAL (Revenue)         0
COMMERICIAL (Sales)           0
COMMERICIAL (Customers)       0
COMMERICIAL (Price)           0
INDUSTRIAL (Revenue)          0
INDUSTRIAL (Sales)            0
INDUSTRIAL (Customers)        0
INDUSTRIAL (Price)            0
TRANSPORTATION (Revenue)      0
TRANSPORTATION (Sales)        0
TRANSPORTATION (Customers)    0
TRANSPORTATION (Price)        0
TOTAL (Revenue)               0
TOTAL (Sales)                 0
TOTAL (Customers)             0
TOTAL (Price)                 0
Date                          0
STATE_NAME                    0
dtype: int64

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20502 entries, 0 to 20501
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Unnamed: 0                  20502 non-null  int64  
 1   State                       20502 non-null  object 
 2   RESIDENTIAL (Revenue)       20502 non-null  object 
 3   RESIDENTIAL (Sales)         20502 non-null  object 
 4   RESIDENTIAL (Customers)     20502 non-null  object 
 5   RESIDENTIAL (Price)         20502 non-null  float64
 6   COMMERICIAL (Revenue)       20502 non-null  object 
 7   COMMERICIAL (Sales)         20502 non-null  object 
 8   COMMERICIAL (Customers)     20502 non-null  object 
 9   COMMERICIAL (Price)         20502 non-null  object 
 10  INDUSTRIAL (Revenue)        20502 non-null  object 
 11  INDUSTRIAL (Sales)          20502 non-null  object 
 12  INDUSTRIAL (Customers)      20502 non-null  object 
 13  INDUSTRIAL (Price)          205

In [114]:
import pandas as pd
import numpy as np


numeric_columns = df.columns.difference(['State','STATE_NAME', 'Date'])


df[numeric_columns] = df[numeric_columns].replace({',': '', '\.': '', ' ': ''}, regex=True)
df[numeric_columns] = df[numeric_columns].replace('', np.nan).astype(float)

# Create the 'Unit Generated' column
df['Unit Generated'] = df['TOTAL (Sales)'] / df['TOTAL (Price)']



In [54]:
df.head()

Unnamed: 0 State  RESIDENTIAL (Revenue)  RESIDENTIAL (Sales)  \
0         0.0    AK                36139.0             222134.0   
1         1.0    AL               276749.0            2818685.0   
2         2.0    AR               125142.0            1448385.0   
3         3.0    AZ               221494.0            2284523.0   
4         4.0    CA              1187506.0            8064319.0   

   RESIDENTIAL (Customers)  RESIDENTIAL (Price)  COMMERICIAL (Revenue)  \
0                 270456.0                16.27                36037.0   
1                2132487.0                 9.82               166408.0   
2                1317936.0                 8.64                67431.0   
3                2563532.0                 9.70               185595.0   
4               13352574.0                14.73              1134995.0   

   COMMERICIAL (Sales)  COMMERICIAL (Customers)  COMMERICIAL (Price)  ...  \
0             256270.0                  47334.0               1406.0  ...   
1            1679995.0                 351480.0                991.0  ...   
2             889269.0                 178058.0                758.0  ...   
3            2156813.0                 299584.0                861.0  ...   
4           10031863.0                1858683.0               1131.0  ...   

   TRANSPORTATION (Sales)  TRANSPORTATION (Customers)  TRANSPORTATION (Price)  \
0                     0.0                         0.0                     0.0   
1                     0.0                         0.0                     0.0   
2                    39.0                         2.0                  1026.0   
3                     0.0                         0.0                     0.0   
4                 68156.0                       495.0                   824.0   

   TOTAL (Revenue)  TOTAL (Sales)  TOTAL (Customers)  TOTAL (Price)  \
0          87685.0       594006.0           319030.0          14.76   
1         603260.0      6964800.0          2493148.0           8.66   
2         268283.0      3621183.0          1525714.0           7.41   
3         462196.0      5341392.0          2870201.0           8.65   
4        2679317.0     21942679.0         15286390.0          12.21   

         Date  STATE_NAME  Unit Generated  
0  2009-12-01      Alaska    4.024431e+04  
1  2009-12-01     Alabama    8.042494e+05  
2  2009-12-01    Arkansas    4.886887e+05  
3  2009-12-01     Arizona    6.175020e+05  
4  2009-12-01  California    1.797107e+06  

[5 rows x 25 columns]

In [116]:
df['Unit_Generated']=df['Unit Generated'].round(2)
df.head(2)

Unnamed: 0 State  RESIDENTIAL (Revenue)  RESIDENTIAL (Sales)  \
0         0.0    AK                36139.0             222134.0   
1         1.0    AL               276749.0            2818685.0   

   RESIDENTIAL (Customers)  RESIDENTIAL (Price)  COMMERICIAL (Revenue)  \
0                 270456.0                16.27                36037.0   
1                2132487.0                 9.82               166408.0   

   COMMERICIAL (Sales)  COMMERICIAL (Customers)  COMMERICIAL (Price)  ...  \
0             256270.0                  47334.0               1406.0  ...   
1            1679995.0                 351480.0                991.0  ...   

   TOTAL (Sales)  TOTAL (Customers)  TOTAL (Price)       Date  STATE_NAME  \
0       594006.0           319030.0          14.76 2009-12-01      Alaska   
1      6964800.0          2493148.0           8.66 2009-12-01     Alabama   

   Unit Generated        Day  Month  Year  Unit_Generated  
0    40244.308943 2009-12-01     12  2009        40244.31  
1   804249.422633 2009-12-01     12  2009       804249.42  

[2 rows x 29 columns]

In [117]:
df['Date'] = pd.to_datetime(df['Date'])
df['Day']=pd.to_datetime(df['Date'])
df['Month'] = df['Date'].dt.month
df['Year'] = df['Date'].dt.year

# Group by 'Month' and sum the 'Unit Generated' column for all states
df1 = df.groupby(['Year', 'Month','Day'])['Unit_Generated'].sum().reset_index()
df1['date']=pd.to_datetime(df['Date'])

In [57]:
df.head(8)

Unnamed: 0 State  RESIDENTIAL (Revenue)  RESIDENTIAL (Sales)  \
0         0.0    AK                36139.0             222134.0   
1         1.0    AL               276749.0            2818685.0   
2         2.0    AR               125142.0            1448385.0   
3         3.0    AZ               221494.0            2284523.0   
4         4.0    CA              1187506.0            8064319.0   
5         5.0    CO               171793.0            1684839.0   
6         6.0    CT               229590.0            1152773.0   
7         7.0    DC                23288.0             173326.0   

   RESIDENTIAL (Customers)  RESIDENTIAL (Price)  COMMERICIAL (Revenue)  \
0                 270456.0                16.27                36037.0   
1                2132487.0                 9.82               166408.0   
2                1317936.0                 8.64                67431.0   
3                2563532.0                 9.70               185595.0   
4               13352574.0                14.73              1134995.0   
5                2118365.0                10.20               136343.0   
6                1451134.0                19.92               179110.0   
7                 226030.0                13.44               100769.0   

   COMMERICIAL (Sales)  COMMERICIAL (Customers)  COMMERICIAL (Price)  ...  \
0             256270.0                  47334.0               1406.0  ...   
1            1679995.0                 351480.0                991.0  ...   
2             889269.0                 178058.0                758.0  ...   
3            2156813.0                 299584.0                861.0  ...   
4           10031863.0                1858683.0               1131.0  ...   
5            1662573.0                 349140.0                 82.0  ...   
6            1083852.0                 153814.0               1653.0  ...   
7             765246.0                  26430.0               1317.0  ...   

   TOTAL (Sales)  TOTAL (Customers)  TOTAL (Price)       Date  \
0       594006.0           319030.0          14.76 2009-12-01   
1      6964800.0          2493148.0           8.66 2009-12-01   
2      3621183.0          1525714.0           7.41 2009-12-01   
3      5341392.0          2870201.0           8.65 2009-12-01   
4     21942679.0         15286390.0          12.21 2009-12-01   
5      4470019.0          2482197.0           8.44 2009-12-01   
6      2542608.0          1609884.0          17.77 2009-12-01   
7       986990.0           252462.0          13.07 2009-12-01   

             STATE_NAME  Unit Generated  Unit_Generated        Day  Month  \
0                Alaska    4.024431e+04        40244.31 2009-12-01     12   
1               Alabama    8.042494e+05       804249.42 2009-12-01     12   
2              Arkansas    4.886887e+05       488688.66 2009-12-01     12   
3               Arizona    6.175020e+05       617501.97 2009-12-01     12   
4            California    1.797107e+06      1797107.21 2009-12-01     12   
5              Colorado    5.296231e+05       529623.10 2009-12-01     12   
6           Connecticut    1.430843e+05       143084.30 2009-12-01     12   
7  District of Columbia    7.551568e+04        75515.68 2009-12-01     12   

   Year  
0  2009  
1  2009  
2  2009  
3  2009  
4  2009  
5  2009  
6  2009  
7  2009  

[8 rows x 29 columns]

Comparison of Two Groups: t-test

Null Hypothesis (H0): There is no significant difference in the mean 'Unit Generated' between two groups

Alternative Hypothesis (H1): There is a significant difference in the mean 'Unit Generated' between the groups.

In [87]:
df1.head(5)

Year  Month        Day  Unit_Generated       date
0  1990      1 1990-01-01     40273223.12 2009-12-01
1  1990      2 1990-02-01     35604480.01 2009-12-01
2  1990      3 1990-03-01     35469737.05 2009-12-01
3  1990      4 1990-04-01     34127677.04 2009-12-01
4  1990      5 1990-05-01     34243910.10 2009-12-01

In [59]:
from scipy.stats import ttest_ind

# Assume df1 is the DataFrame with 'Unit Generated' grouped by year and month
group1 = df1[(df1['Year'] >= 2009) & (df1['Year'] <= 2016)]['Unit_Generated']
group2 = df1[(df1['Year'] >= 2016) & (df1['Year'] <= 2023)]['Unit_Generated']

t_stat, p_value = ttest_ind(group1, group2)
print(f"T-statistic: {t_stat}, p-value: {p_value}")

T-statistic: 4.696607939665177, p-value: 5.160037440243047e-06


t-statistic of approximately 5.09 and a very small p-value (4.696607939665177). The p-value is below conventional significance levels (e.g., 0.05), suggesting that you would reject the null hypothesis.

In [60]:
df['STATE_NAME'].unique()

array(['Alaska', 'Alabama', 'Arkansas', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'District of Columbia', 'Delaware',
       'Florida', 'Georgia', 'Hawaii', 'Iowa', 'Idaho', 'Illinois',
       'Indiana', 'Kansas', 'Kentucky', 'Louisiana', 'Massachusetts',
       'Maryland', 'Maine', 'Michigan', 'Minnesota', 'Missouri',
       'Mississippi', 'Montana', 'North Carolina', 'North Dakota',
       'Nebraska', 'New Hampshire', 'New Jersey', 'New Mexico', 'Nevada',
       'New York', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Virginia', 'Vermont', 'Washington', 'Wisconsin',
       'West Virginia', 'Wyoming'], dtype=object)

In [61]:
df3 = df.groupby(['STATE_NAME', 'Year'])['Unit Generated'].sum().reset_index()

In [62]:
print(df3[df3['STATE_NAME'] == 'New York'])

     STATE_NAME  Year  Unit Generated
1088   New York  1990    1.382564e+07
1089   New York  1991    1.325913e+07
1090   New York  1992    1.269577e+07
1091   New York  1993    1.221222e+07
1092   New York  1994    1.207021e+07
1093   New York  1995    1.184426e+07
1094   New York  1996    1.186036e+07
1095   New York  1997    1.189269e+07
1096   New York  1998    1.265097e+07
1097   New York  1999    1.372566e+07
1098   New York  2000    1.257442e+07
1099   New York  2001    1.254096e+07
1100   New York  2002    1.325978e+07
1101   New York  2003    1.162439e+07
1102   New York  2004    1.160266e+07
1103   New York  2005    1.084371e+07
1104   New York  2006    9.345329e+06
1105   New York  2007    9.752039e+06
1106   New York  2008    8.819149e+06
1107   New York  2009    9.093876e+06
1108   New York  2010    8.836509e+06
1109   New York  2011    9.091678e+06
1110   New York  2012    9.474405e+06
1111   New York  2013    9.607345e+06
1112   New York  2014    9.088190e+06
1113   New Y

In [63]:
import pandas as pd
from scipy.stats import ttest_ind

# Get unique state names
states = df['STATE_NAME'].unique()

# Initialize lists to store the results
results = []

# Perform t-test for each pair of states
for i in range(len(states)):
    for j in range(i + 1, len(states)):
        state1 = states[i]
        state2 = states[j]

        # Select data for state1 and state2 from the respective time periods
        group_2009_to_2016_state1 = df[(df['STATE_NAME'] == state1) & (df['Year'] >= 2009) & (df['Year'] <= 2016)]['Unit_Generated']
        group_2016_to_2023_state1 = df[(df['STATE_NAME'] == state1) & (df['Year'] >= 2016) & (df['Year'] <= 2023)]['Unit_Generated']

        group_2009_to_2016_state2 = df[(df['STATE_NAME'] == state2) & (df['Year'] >= 2009) & (df['Year'] <= 2016)]['Unit_Generated']
        group_2016_to_2023_state2 = df[(df['STATE_NAME'] == state2) & (df['Year'] >= 2016) & (df['Year'] <= 2023)]['Unit_Generated']

        # Perform t-test for the two pairs of states
        t_statistic_state1, p_value_state1 = ttest_ind(group_2009_to_2016_state1, group_2016_to_2023_state1, equal_var=False)
        t_statistic_state2, p_value_state2 = ttest_ind(group_2009_to_2016_state2, group_2016_to_2023_state2, equal_var=False)

        # Store the results in a list
        results.append({
            'State 1': state1,
            'State 2': state2,
            'P-value': p_value_state1,
            'T-statistic': t_statistic_state1
        })
# Create a DataFrame from the results list
results_df = pd.DataFrame(results)

In [64]:
# Print the DataFrame
results_df

State 1        State 2       P-value  T-statistic
0            Alaska        Alabama  5.686674e-25    12.147511
1            Alaska       Arkansas  5.686674e-25    12.147511
2            Alaska        Arizona  5.686674e-25    12.147511
3            Alaska     California  5.686674e-25    12.147511
4            Alaska       Colorado  5.686674e-25    12.147511
...             ...            ...           ...          ...
1270     Washington  West Virginia  6.540547e-24    11.670680
1271     Washington        Wyoming  6.540547e-24    11.670680
1272      Wisconsin  West Virginia  6.020682e-07     5.175932
1273      Wisconsin        Wyoming  6.020682e-07     5.175932
1274  West Virginia        Wyoming  1.226232e-08     5.979392

[1275 rows x 4 columns]

In [65]:
print(i)

50


In [66]:
len(states)

51

In [67]:
from math import factorial
total_states = len(states)
combinations = factorial(total_states) / (factorial(2) * factorial(total_states - 2))

In [68]:
combinations

1275.0

In [69]:
# Filter rows where p_value_state1 is less than or equal to 0.05
filtered_results = results_df[results_df['P-value'] <= 0.05]

In [70]:
# Count the number of rows in the filtered results
count_filtered_results = len(filtered_results)

# Print the count
print("Number of rows in filtered results:", count_filtered_results)


Number of rows in filtered results: 949


In [71]:
filtered_results

State 1        State 2       P-value  T-statistic
0            Alaska        Alabama  5.686674e-25    12.147511
1            Alaska       Arkansas  5.686674e-25    12.147511
2            Alaska        Arizona  5.686674e-25    12.147511
3            Alaska     California  5.686674e-25    12.147511
4            Alaska       Colorado  5.686674e-25    12.147511
...             ...            ...           ...          ...
1270     Washington  West Virginia  6.540547e-24    11.670680
1271     Washington        Wyoming  6.540547e-24    11.670680
1272      Wisconsin  West Virginia  6.020682e-07     5.175932
1273      Wisconsin        Wyoming  6.020682e-07     5.175932
1274  West Virginia        Wyoming  1.226232e-08     5.979392

[949 rows x 4 columns]

So, to reject the null hypothesis:

The p-value should be less than or equal to the chosen significance level (usually 0.05).
The absolute value of the t-statistic should be large.

In [ ]:
import pandas as pd

# Assuming model_auto.predict returns a list or array of predictions
predictions = model_auto.predict(n_periods=len(df1_test[321:402]))

# Creating a DataFrame with predictions and setting the index
df_auto_pred = pd.DataFrame(predictions, index=df1_test[321:402].index)
